Add multiple writer threads to improve throughput

rajubairishetti opened the pull request:
Refer https://jira.corp.inmobi.com/browse/SCRIBE-37

rajubairishetti committed:
Scribe-37: find least sized store queue and add message to that

rajubairishetti committed:
create multiple store queues for categories

rajubairishetti committed:
pass storeQueue/storeThread name to the storeQueue

rajubairishetti committed:
change the signature of few methods to accept store thread name

rajubairishetti committed:
append thread to the subdirectory(i.e. hostname), create a subdirectory with thread_name and change the singature of store methods to pass thread name

rajubairishetti committed:
accept the messages even if atleast one queue has enough space and make num_store_threads config per category

rajubairishetti committed:
add the messages to least sized queue in case if it is multi writer threaded

rajubairishetti committed:
add function prototype in scribe_server.h header file

rajubairishetti committed:
check for the storequeues of a msg category instead of all

rajubairishetti committed:
create single thread for model categories and fix todos

rajubairishetti committed:
remove all todos

rajubairishetti committed:
remove the log statements

rajubairishetti committed:
remove commented code and remve unused variables

rajubairishetti committed:
fix typo and fix condition to find prefix

rajubairishetti committed:
use '_' instead of '-' in collector name

rajubairishetti commented on src/scribe_server.cpp:
```diff
@@ -388,38 +389,39 @@ bool scribeHandler::throttleRequest(const vector<LogEntry>&  messages) {
   }
-
+  if (throttle) {
+    incCounter(category, "denied for queue size");
```
Will log the information of throttle as part of [SCRIBE-30](https://jira.corp.inmobi.com/browse/SCRIBE-30)

satish-mittal commented on src/conf.cpp:
```diff
@@ -119,7 +119,8 @@ bool StoreConf::getString(const string& stringName,
       || stringName == "category"
-      || stringName == "categories") {
+      || stringName == "categories"
+      || stringName == "num_store_threads") {
```
why is this needed?

rajubairishetti replied:
This is required. Otherwise this property will be inherited to the stores (if we do not specify the num_store_threads conf) which are configured below this store-category in the scribe conf.

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -388,38 +389,39 @@ bool scribeHandler::throttleRequest(const vector<LogEntry>&  messages) {
+    } else {
+      unsigned long long size = (*store_iter)->getSize();
+      if (size <= maxQueueSize) {
+        throttle = false;
```
Scribe changes to improve throttling logic is tracked in a separate jira SCRIBE-29. Can we undo the throttling change from this jira and have it together in a separate pull request for SCRIBE-29?

rajubairishetti replied:
yeah..sure.. I will revert this commit

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -439,7 +441,12 @@ shared_ptr<store_list_t> scribeHandler::createNewCategory(
           store_iter != pstores->end(); ++store_iter) {
-        createCategoryFromModel(category, *store_iter);
+        // currently we are not supporting multiple threads for the model categories.
+        // specify new thread name for each new store queue
```
modify comments to be more clear.

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -439,7 +441,12 @@ shared_ptr<store_list_t> scribeHandler::createNewCategory(
+        // specify new thread name for each new store queue
+        ostringstream ostr;
+        ostr << "";
+        std::string thread_name = ostr.str();
```
you may want to instead define a static string empty_str as a global variable and use it.

satish-mittal replied:
Actually you can define thread_name as an optional parameter in scribe_server.h, such that it has a default value of "" in function declaration in .h file itself. That way you won't need to pass on default value of "" at each place.

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -460,7 +467,12 @@ shared_ptr<store_list_t> scribeHandler::createNewCategory(
         store_iter != defaultStores.end(); ++store_iter) {
-      createCategoryFromModel(category, *store_iter);
+      // currently we are not supporting multiple threads for the model categories.
+      // Specify a new threadName for each storeQueue thread in case of multi-thread
```
same as above.

rajubairishetti committed:
revert 441c532c24 commit and fix conflicts

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -500,6 +512,41 @@ void scribeHandler::addMessage(
 }
 
+// Add this message to every least sized store in list
+void scribeHandler::addMessageToLeastSizedQueue(
```
Instead of exposing a new method, can we merge this logic with existing addMessage()?

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -500,6 +512,41 @@ void scribeHandler::addMessage(
+  if (numstores) {
+	// find the least sized store queue add the
+	int minIndex = 0;
+	size_t min_queue_size = (*store_list)[0]->getSize();
```
The current scribe code iterates through std::vector using iterator mechanism. E.g. store_list_t::iterator store_iter...

Let us continue with that style for having seamless code and behavior, instead of mixing it with index based iteration, e.g. (*store_list)[i].

rajubairishetti committed:
define an empty_string variable and fix comments

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -873,14 +923,35 @@ bool scribeHandler::configureStore(pStoreConf store_conf, int *numstores) {
+        shared_ptr<StoreQueue> result =
+            configureStoreCategory(store_conf, category_list[0], model, thread_name);
+        if (result == NULL) {
+          LOG_OPER("Unable to create store queue [%s] for [%s] catgeory", thread_name.c_str(), category_list[0].c_str());
```
fix typo for 'category'

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -927,6 +1002,7 @@ shared_ptr<StoreQueue> scribeHandler::configureStoreCategory(
   pStoreConf store_conf,                       //configuration for store
   const string &category,                      //category name
   const boost::shared_ptr<StoreQueue> &model,  //model to use (optional)
+  string& thread_name,                         //storeQueue name
```
Typically new arguments are added to the end of an existing signature (i.e. after category_list). Also declare it const string&.

rajubairishetti committed:
remove addMessageToLeastSizedQueue method
add the required changes in existing addMessag method
Iterate the store_list through iterator

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -873,14 +923,35 @@ bool scribeHandler::configureStore(pStoreConf store_conf, int *numstores) {
+
+    unsigned long int num_store_threads = -1;
+    if (!store_conf->getUnsigned("num_store_threads", num_store_threads)
+      || is_default_category || is_prefix_category || (num_store_threads <= 0)) {
```
check for num_store_threads <= 1

satish-mittal commented on src/scribe_server.h:
```diff
@@ -130,11 +130,13 @@ class scribeHandler : virtual public scribe::thrift::scribeIf,
   bool createCategoryFromModel(const std::string &category,
-                               const boost::shared_ptr<StoreQueue> &model);
+                               const boost::shared_ptr<StoreQueue> &model,
+                               std::string& thread_name);
```
make it const and default value as empty str

satish-mittal commented on src/store.cpp:
```diff
@@ -81,34 +81,34 @@ bool shouldSendDummy(boost::shared_ptr<logentry_vector_t> messages) {
+    return shared_ptr<Store>(new BufferStore(storeq,category, thread_name, multi_category));
   } else if (0 == type.compare("network")) {
-    return shared_ptr<Store>(new NetworkStore(storeq, category,
+    return shared_ptr<Store>(new NetworkStore(storeq, category, thread_name,
```
Currently we are adding this feature for only file store. Hence we should pass thread_name arg only for file and buffer.

satish-mittal commented on src/store.cpp:
```diff
@@ -117,9 +117,11 @@ Store::createStore(StoreQueue* storeq, const string& type,
 
 Store::Store(StoreQueue* storeq,
              const string& category,
+             string& thread_name,
```
declare it const

satish-mittal commented on src/store.cpp:
```diff
@@ -250,6 +252,9 @@ void FileStoreBase::configure(pStoreConf configuration, pStoreConf parent) {
     filePath += "/" + subDirectory;
   }
 
+  if (!isPrimary &&  !threadName.empty() && subDirectory.empty()) {
```
We can remove the check for subDirectory.empty since we are not using it to build spooling path.

satish-mittal commented on src/store.cpp:
```diff
@@ -622,6 +631,10 @@ void FileStoreBase::setHostNameSubDir() {
   }
+  // append threadName to the subdirectory
+  if (!threadName.empty()) {
+    subDirectory = hoststring + "_" +threadName;
```
missing space after +

satish-mittal commented on src/store.cpp:
```diff
@@ -750,6 +763,10 @@ bool FileStore::openInternal(bool incrementFilename, struct tm* current_time) {
       success = writeFile->createDirectory(filePath);
     }
 
+    if (!isPrimary && !threadName.empty()) {
```
check for success as the first condition inside if

satish-mittal commented on src/store.h:
```diff
@@ -255,11 +257,11 @@ class FileStore : public FileStoreBase {
  public:
   ThriftFileStore(StoreQueue* storeq,
-                  const std::string& category,
+                  const std::string& category, std::string& thread_name,
```
Undo all the signature changes to add thread_name in unwanted stores.

satish-mittal commented on src/store.cpp:
```diff
@@ -1480,12 +1499,12 @@ void BufferStore::configure(pStoreConf configuration, pStoreConf parent) {
-      primaryStore = createStore(storeQueue, type, categoryHandled, false,
+      primaryStore = createStore(storeQueue, type, categoryHandled, threadName, false,
                                   multiCategory);
-      primaryStore->configure(primary_store_conf, storeConf);
```
do we need the change the order of these 2 calls?

rajubairishetti committed:
fix review comments

rajubairishetti committed:
fix condition

rajubairishetti committed:
fix min_store_queue

rajubairishetti committed:
change store method signature

rajubairishetti committed:
change method signature and fix method invocations in store.cpp

rajubairishetti committed:
fix method singature

rajubairishetti committed:
fix indentation

rajubairishetti committed:
change the comment

rajubairishetti committed:
use same string const across multiple headers

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -439,6 +439,7 @@ shared_ptr<store_list_t> scribeHandler::createNewCategory(
       shared_ptr<store_list_t> pstores = cat_prefix_iter->second;
       for (store_list_t::iterator store_iter = pstores->begin();
           store_iter != pstores->end(); ++store_iter) {
+        // currently we are not supporting multiple threads for prefix model
```
Can you enhance it to: "Currently we are not supporting multiple threads for prefix model. Hence not passing any thread name".

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -481,19 +482,46 @@ void scribeHandler::addMessage(
+    if (!(*store_iter)->getThreadName().empty()) {
+      isMultiThreaded = true;
+    }
+    break;
```
you should move this break to inside the if condition; else you immediately break out of for loop after one iteration.

rajubairishetti replied:
Here, we just wanted to know whether the category store is configured with multiple threads or not. It should come out from the loop if thread name is not defined for store queue thread.

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -481,19 +482,46 @@ void scribeHandler::addMessage(
+    break;
+  }
+
+  // Find least sized queue in case if category store configured with multiple threads.
```
Rephrase the comment to: 
If category store is configured with multiple threads, then add message to the least sized queue. Else simply add it to all stores in the queue.

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -481,19 +482,46 @@ void scribeHandler::addMessage(
+      (*store_iter)->addMessage(ptr);
+    } else {
+	  if (min_queue_size > (*store_iter)->getSize()) {
+        min_queue_size = (*store_iter)->getSize();
```
Fix indentation of these 2 lines inside if condition

rajubairishetti committed:
fix review comments

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -481,19 +484,46 @@ void scribeHandler::addMessage(
-    (*store_iter)->addMessage(ptr);
+      (*store_iter)->addMessage(ptr);
+    } else {
+      if (min_queue_size > (*store_iter)->getSize()) {
```
As we discussed, please re-organize the code. Also since min_queue_size is -1 initially, we will never enter this if condition.

satish-mittal commented on src/scribe_server.h:
```diff
@@ -143,6 +145,8 @@ class scribeHandler : virtual public scribe::thrift::scribeIf,
     createNewCategory(const std::string& category);
   void addMessage(const scribe::thrift::LogEntry& entry,
                   const boost::shared_ptr<store_list_t>& store_list);
+  void addMessageToLeastSizedQueue(const scribe::thrift::LogEntry& entry,
```
Why is this method still there?

satish-mittal commented on src/store.h:
```diff
@@ -391,10 +399,11 @@ class NetworkStore : public Store {
   NetworkStore(StoreQueue* storeq,
                const std::string& category,
-               bool multi_category);
+               bool multi_category, const std::string& thread_name=empty_string);
```
why do we need thread name in this ctor?

rajubairishetti committed:
fix review comments

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -330,7 +330,8 @@ const char* scribeHandler::statusAsString(fb_status status) {
 bool scribeHandler::createCategoryFromModel(
-  const string &category, const boost::shared_ptr<StoreQueue> &model) {
+  const string &category, const boost::shared_ptr<StoreQueue> &model,
+  const string&  thread_name) {
```
Since we have decided that model categories will not support multiple threads as of now, we don't need this signature change.

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -439,6 +439,8 @@ shared_ptr<store_list_t> scribeHandler::createNewCategory(
       for (store_list_t::iterator store_iter = pstores->begin();
           store_iter != pstores->end(); ++store_iter) {
+        // Currently we are not supporting multiple threads for prefix model.
+        // Hence not passing any thread name
```
Remove this comment since not needed.

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -455,11 +457,12 @@ shared_ptr<store_list_t> scribeHandler::createNewCategory(
     for (store_list_t::iterator store_iter = defaultStores.begin();
         store_iter != defaultStores.end(); ++store_iter) {
+      // currently we are not supporting multiple threads for default model.
+      // Hence not passing any thread name
```
Remove this comment.

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -873,14 +903,32 @@ bool scribeHandler::configureStore(pStoreConf store_conf, int *numstores) {
+        return false;
+      }
+    } else {
+      for (std::size_t i = 0; i < num_store_threads; i++) {
```
Log a message that prints the value of num_store_threads config.

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -873,14 +903,32 @@ bool scribeHandler::configureStore(pStoreConf store_conf, int *numstores) {
+        ostringstream ostr;
+        ostr << "thread_" << i;
+        const std::string thread_name = ostr.str();
+        LOG_OPER("Store Queue name [%s] for [%s] category", thread_name.c_str(), category_list[0].c_str());
```
Move this log message down to the condition when store queue is configured.

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -962,7 +1010,7 @@ shared_ptr<StoreQueue> scribeHandler::configureStoreCategory(
       // Create a copy of the model if we want a new thread per category
       if (newThreadPerCategory && !is_default && !is_prefix_category) {
-        pstore = shared_ptr<StoreQueue>(new StoreQueue(model, category));
+        pstore = shared_ptr<StoreQueue>(new StoreQueue(model, category, thread_name));
```
Remove this code since it is the case of model based store creation.

satish-mittal commented on src/store.h:
```diff
@@ -55,13 +54,16 @@ class Store {
   Store(StoreQueue* storeq, const std::string& category,
-        const std::string &type, bool multi_category = false);
+        const std::string &type, bool multi_category = false,
+        const std::string& thread_name = empty_string);
```
we don't need to have thread_name in base class if we can keep thread_name inside buffer store itself.

satish-mittal commented on src/store.h:
```diff
@@ -55,13 +54,16 @@ class Store {
 
-  virtual boost::shared_ptr<Store> copy(const std::string &category) = 0;
+  virtual boost::shared_ptr<Store> copy(const std::string &category,
+		  const std::string& thread_name = empty_string) = 0;
```
This copy is not needed since this method is called only while creating store queue from a model.

rajubairishetti committed:
remove generic things from method signature

rajubairishetti committed:
fix method signature

rajubairishetti committed:
add a log statement

satish-mittal commented on src/scribe_server.cpp:
```diff
@@ -873,14 +900,34 @@ bool scribeHandler::configureStore(pStoreConf store_conf, int *numstores) {
+        return false;
+      }
+    } else {
+      LOG_OPER("Configuring [%lu] store queues for [%s] ", num_store_threads, category_list[0].c_str());
```
category_list[0].c_str() is being repeated multiple times. You could keep a local var for it.

satish-mittal commented on src/store.cpp:
```diff
@@ -644,8 +651,9 @@ void FileStoreBase::auditFileClosed() {
-  : FileStoreBase(storeq, category, "file", multi_category),
+                     bool multi_category, bool is_buffer_file,
+                     const string &thread_name)
+  : FileStoreBase(storeq, category, "file", multi_category, thread_name),
```
In addition to stream level counters, can we introduce a stream_thread level counter and increment it? That will help us get more insights into the write performance at thread level.

satish-mittal commented:
You need to run benchmark at local/merge level to ensure that these scribe changes work fine with current conduit worker (i.e. no changes required in conduit worker).

rajubairishetti committed:
have a char pointer local varibale

rajubairishetti committed:
maintain a hdfs_wrote_messages/bytes counter per thread also

satish-mittal commented:
+1
Please run benchmark tests for merge also.
